File: csharp/simple/src/Core/GameState.cs
namespace Core{
    public class GameState {
        // 全てイミュータブルなので、getterでのディープコピーは必要ない。
        public int? LastInputNumber {get; init;} = null;
        public int CurrentAttempt {get; init;} = 0;
        public bool IsSuccess {get; init;} = false;
        // 大した容量でもないし頻度も低いので再生成しても問題ない。
        public static GameState UpdateState(int input, GameState gameState, GameSetting gameSetting) {
            return new GameState {
                LastInputNumber = input,
                CurrentAttempt = gameState.CurrentAttempt + 1,
                IsSuccess = (gameSetting.SecretNumber == input) ? true : false
            };
        }
    }
}

File: cpp/simple/src/core/GameState.cpp
#include "GameState.hpp"

GameState GameState::update_state(
  int input, const GameState & game_state, const GameSetting & game_setting) const
{
  return GameState(
    input, game_state.current_attempt_ + 1, game_setting.SECRET_NUM == input ? true : false);
}
